lib/commands/generate-json.js

```diff
@@ -0,0 +1,76 @@
+var fs              = require('fs');
+var path            = require('path');
+var mkdirp          = require('mkdirp');
+var Command         = require('../models/command');
+var JsonGenerator   = require('../utilities/json-generator');
+
+module.exports = Command.extend({
+  name: 'generate-json',
+  description: 'Generates a cache describing all commands and blueprints available for the command line',
+  aliases: [],
+  skipHelp: true,
+  works: 'everywhere',
+
+  availableOptions: [
+    { name: 'silent', type: Boolean, default: true, aliases: ['s'] },
+  ],
+
+  anonymousOptions: [],
+
+  cacheDirectory: '/node_modules/ember-cli',
+  cacheFileName: 'commands-help.json',
+
+  run: function(commandOptions, rawArgs) {
+
+    commandOptions.json = true;
+    var json = this._getJson(commandOptions, rawArgs);
+    var cache = JSON.stringify(json, function(key, value) {
+      var path = require('path');
+      // build command has a recursive property
+      if (value === path) {
+        return 'path';
+      }
+      return value;
+    }, 2);
+
+    if (rawArgs.length === 0) {
+      this._cacheJson(cache);
+    }
+
+    return cache;
+  },
+
+  _getJson: function(commandOptions, rawArgs) {
+    var generator = new JsonGenerator({
+      ui: this.ui,
+      project: this.project,
+      commands: this.commands,
+      tasks: this.tasks
+    });
+
+    return generator.generate(commandOptions, rawArgs);
+  },
+
+  _cacheJson: function(json) {
+
+    var directoryPath = path.join(__dirname, '../..')
```

Comment:
no i miss-pasted there: it would be
`this.project.root: /Users/lazybensch/src/opensource/ember-cli-filter-by-query` (which would still be correct. when i join it with 'node_modules/ember-cli' but since this directory is just a symlink to `/Users/lazybensch/src/opensource/ember-cli` it has the same effect as path.join(__dirname, '../..')

so still, both versions work inside a linked project
both versions work inside a not-linked project
only __dirname works when outside a project